pkg/wasm/cache.go

```diff
@@ -0,0 +1,184 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"istio.io/pkg/log"
+)
+
+var wasmLog = log.RegisterScope("wasm", "", 0)
+
+const (
+	// DefaultWasmModulePurgeInteval is the default interval for periodic stale Wasm module clean up.
+	DefaultWasmModulePurgeInteval = 10 * time.Minute
+
+	// DefaultWasmModuleExpiry is the default duration for least recently touched Wasm module to become stale.
+	DefaultWasmModuleExpiry = 24 * time.Hour
+)
+
+// Cache models a Wasm module cache.
+// TODO(bianpengyuan): add metrics to the wasm package.
+type Cache interface {
+	Get(url, checksum string, timeout *time.Duration) (string, error)
```

Comment:
timeout as a pointer seems a bit strange?

Reply:
I was trying to find a way to represent `no timeout is configured by user`. I don't see a handy way to check if a duration is zero in golang so I make it a pointer. I also feel checking zero duration is better than pointer. Is the common practice is just to check `if duration == 0 * time.Second()`?

Reply:
I think you can just do `dur == 0`, duration is a alias for int64 if I recall right. I don't know if its common practice, but it works :slightly_smiling_face: .

more common practice would probably be to use a context.Context so user can add a cancel or timeout, but we don't do it too much throughout our codebase

Reply:
Thanks! updated.